### PR TITLE
arch/arm/stm32: Basic STM32F7 family support

### PR DESCRIPTION
**Pull request by pic16f887**
Related to issue #6981.
Added support for STM32F746XG subfamily.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

### CODE REVIEW

File: arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#
+
+if SOC_SERIES_STM32F7X
+
+source "arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f7*"
+
+config SOC_SERIES
+	default stm32f7
+
+if USB
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series]: Remove USB as it is not yet tested/implemented on this series (limited to F4 for now)

File: arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	default 0x82
+
+config CDC_ACM_OUT_EP_ADDR
+	default 0x03
+
+endif # USB_CDC_ACM
+
+endif #USB
+
+if I2C && (I2C_1 || I2C_2 || I2C_3 || I2C_4)
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series]: Remove I2C, not yet tested/supported on F7 series

File: arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f746xg [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	default 97
+
+if GPIO_STM32
+
+config GPIO_STM32_PORTI
+	default y
+
+endif # GPIO_STM32
+
+if ENTROPY_GENERATOR
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f746xg]: Remove for now. You could add it in a follow up PR.

File: arch/arm/soc/st_stm32/stm32f7/dts.fixup [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+#define CONFIG_UART_STM32_PORT_3_NAME			ST_STM32_USART_40004800_LABEL
+#define PORT_3_IRQ								ST_STM32_USART_40004800_IRQ_0
+
+#define CONFIG_UART_STM32_PORT_6_NAME			ST_STM32_USART_40011400_LABEL
+#define CONFIG_UART_STM32_PORT_6_BASE_ADDRESS	ST_STM32_USART_40011400_BASE_ADDRESS
+#define CONFIG_UART_STM32_PORT_6_BAUD_RATE		ST_STM32_USART_40011400_CURRENT_SPEED
+#define CONFIG_UART_STM32_PORT_6_IRQ_PRI		ST_STM32_USART_40011400_IRQ_0_PRIORITY
+#define PORT_6_IRQ								ST_STM32_USART_40011400_IRQ_0
+
+#define CONFIG_I2C_1_BASE_ADDRESS				ST_STM32_I2C_V1_40005400_BASE_ADDRESS
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/dts.fixup]: Remove I2C, SPI, USB and flash

File: arch/arm/soc/st_stm32/stm32f7/flash_registers.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2018 Yurii Hamann
+ * Copyright (c) 2016 Linaro Limited.
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/flash_registers.h]: Keep only your copyright, apply on all files

File: arch/arm/soc/st_stm32/stm32f7/soc.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#include <stm32f7xx_ll_rcc.h>
+#include <stm32f7xx_ll_system.h>
+#include <stm32f7xx_ll_spi.h>
+#endif /* CONFIG_CLOCK_CONTROL_STM32_CUBE */
+
+#ifdef CONFIG_SERIAL_HAS_DRIVER
+#include <stm32f7xx_ll_usart.h>
+#endif
+
+#ifdef CONFIG_I2C
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/soc.h]: Remore I2C

File: arch/arm/soc/st_stm32/stm32f7/soc.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+#include <stm32f7xx_ll_i2c.h>
+#endif
+
+#ifdef CONFIG_ENTROPY_STM32_RNG
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/soc.h]: Remove Entropy

File: arch/arm/soc/st_stm32/stm32f7/soc.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+#include <stm32f7xx_ll_rng.h>
+#endif
+
+#ifdef CONFIG_IWDG_STM32
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/soc.h]: Remove

File: arch/arm/soc/st_stm32/stm32f7/soc.c [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * @brief Perform basic hardware initialization at boot.
+ *
+ * This needs to be run from the very beginning.
+ * So the init priority has to be 0 (zero).
+ *
+ * @return 0
+ */
+static int st_STM32F7_init(struct device *arg)
```
Comment (erwango) [inline arch/arm/soc/st_stm32/stm32f7/soc.c]: keep lower case

File: drivers/pinmux/stm32/pinmux_stm32f7.h [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+ */
+
+#ifndef _STM32F7_PINMUX_H_
+#define _STM32F7_PINMUX_H_
+
+/**
+ * @file Header for STM32F7 pin multiplexing helper
+ */
+
+/* Port A */
```
Comment (erwango) [inline drivers/pinmux/stm32/pinmux_stm32f7.h]: Clean everything which is not UART

File: drivers/serial/uart_stm32.c [hunk 11]
```diff
@@ -295,7 +295,7 @@ static int uart_stm32_init(struct device *dev)
 
 	LL_USART_Enable(UartInstance);
 
-#if !defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_SERIES_STM32F1X)
+#if !defined(CONFIG_SOC_SERIES_STM32F4X) && !defined(CONFIG_SOC_SERIES_STM32F7X) && !defined(CONFIG_SOC_SERIES_STM32F1X)
```
Comment (erwango) [inline drivers/serial/uart_stm32.c]: Take care to the 80 char limit

File: dts/arm/st/stm32f7.dtsi [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x40005400 0x400>;
+			interrupts = <31 0>, <32 0>;
+			interrupt-names = "event", "error";
+			status = "disabled";
+			label= "I2C_1";
+		};
+
+		i2c2: i2c@40005800 {
```
Comment (erwango) [inline dts/arm/st/stm32f7.dtsi]: clean up i2c spi and USB

**Review by erwango — CHANGES_REQUESTED** [review]
Hi @pic16f887 ,

Thanks for this work.
Some initial comments on the pull request:
 * focus your work on uart, gpio, exti and clock-control drivers which are required for basic testing (hello_world and blinky). Clean up everything is relates to USB, I2C, SPI, .... It will ease to focus and fasten the review process.
  *Split your work into coherent commit (arch, dts, drivers/clock_control, drivers/uart, ...)
  * Last: please port a F7 board (in a dedicated commit) so the code could at least be compiled

Also, you should add your Signed-Off in the commit message

Thanks again!

File: arch/arm/soc/st_stm32/stm32f7/Kconfig.series [hunk 13]
```diff
@@ -0,0 +1,20 @@
+# Kconfig - ST Microelectronics STM32F7 MCU series
+#
+# Copyright (c) 2018 Yurii Hamann
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config SOC_SERIES_STM32F7X
+	bool "STM32F7x Series MCU"
+	select CPU_CORTEX_M
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f7/Kconfig.series]: You should remove this. CPU_CORTEX_M is selected when you select CPU_CORTEX_M7.

**Comment by pic16f887** [comment]
Hello @erwango ,
I have split the commit into smaller ones as you asked me. I added support for STM32F746G Discovery board. I tested it on on "hello_world" and "blinky" demo.

File: boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst [hunk 14]
```diff
@@ -0,0 +1,200 @@
+.. _stm32f746g_disco_board:
+
+ST STM32F746G Discovery
+#######################
+
+Overview
+********
+
+The discovery kit enables a wide diversity of applications taking benefit from audio, multi-sensor support, graphics, security, video and high-speed connectivity features.
```
Comment (dbkinder) [inline boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst]: Preference is for lines to be <80 characters

Change the ending to read:

   ... security, video, and high-speed connectivity features.  Important board features include:

File: boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+- UART_1_TX : PA9
+- UART_1_RX : PB7
+- UART_3_TX : PC6
+- UART_3_RX : PC7
+- LD1 : PI1
+
+System Clock
+============
+
+STM32F746G System Clock could be driven by internal or external oscillator,
```
Comment (dbkinder) [inline boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst]: lots of missing articles (the, a, an):

    The STM32F746G System Clock can be driven by an internal or external oscillator,
    as well as by the main PLL clock. By default, the System clock is driven by the PLL
    clock at 216MHz, driven by a 25MHz high speed external clock.

File: boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+as well as main PLL clock. By default System clock is driven by PLL clock at 216MHz,
+driven by 25MHz high speed external clock.
+
+Serial Port
+===========
+
+The STM32F746G Discovery kit has up to 8 UARTs. The Zephyr console output is assigned to UART1.
```
Comment (dbkinder) [inline boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst]: I suspect the kit has a fixed number of UARTs, so it's better to say:

    The STM32F746G Discovery kit has 8 UARTs.

File: boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+Default settings are 115200 8N1.
```
Comment (dbkinder) [inline boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst]: ``Default communication settings...``

File: boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+Applications for the ``stm32f746g_disco`` board configuration can be built and
+flashed in the usual way (see :ref:`build_an_application` and
+:ref:`application_run` for more details).
+
+Flashing
+========
+
+STM32F746G Discovery kit includes an ST-LINK/V2 embedded debug tool interface.
+This interface is supported by the openocd version included in Zephyr SDK.
```
Comment (dbkinder) [inline boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst]: ``in the Zephyr``

**Review by dbkinder — CHANGES_REQUESTED** [review]
Thanks for the documentation.  A few fixes...

File: boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst (hunk 16 above)
Comment (pic16f887) [reply boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst]: @dbkinder, 
I am not sure that all 8 UARTs are available. Many ports are occupied by LCD display with touchscreen and other peripherals.

File: arch/arm/soc/st_stm32/stm32f7/soc.h [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+ * Copyright (c) 2018 Yurii Hamann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file SoC configuration macros for the ST STM32F7 family processors.
+ *
+ * Based on reference manual:
+ *   RM0368 Reference manual STM32F401xB/C and STM32F401xD/E
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f7/soc.h]: Please add STM32F7 reference manuals.

File: dts/arm/st/stm32f7.dtsi [hunk 20]
```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2018 Yurii Hamann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <arm/armv7-m.dtsi>
+#include <st/mem.h>
+#include <dt-bindings/clock/stm32_clock.h>
+#include <dt-bindings/i2c/i2c.h>
```
Comment (ydamigos) [inline dts/arm/st/stm32f7.dtsi]: You could remove this.

File: dts/arm/st/stm32f7.dtsi [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+	};
+
+	sram0: memory@20000000 {
+		device_type = "memory";
+		compatible = "mmio-sram";
+		reg = <0x20000000 DT_SRAM_SIZE>;
+	};
+
+	soc {
+		flash-controller@40023c00 {
```
Comment (ydamigos) [inline dts/arm/st/stm32f7.dtsi]: Do we support flash controller for stm32f7 family? I couldn't find it in the flash drivers. If not, please remove this node.

**Review by ydamigos — CHANGES_REQUESTED** [review]

**Review by ydamigos — CHANGES_REQUESTED** [review]
You should cleanup your PR. Please drop the commits that you revert and fix your commits body to make shippable happy.

**Review by dbkinder — CHANGES_REQUESTED** [review]
documentation seems to be missing now?  And there are problems with the commit messages reported by shippable: (https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/14066/5/tests)

**Pull request closed without merging**